packages/cli/src/index.ts

```diff
@@ -146,17 +146,22 @@ async function doBuild(cannonfile: string, settings: string[], opts: any): Promi
 
   let chainId: number | undefined = undefined;
 
-  if (!opts.chainId) {
+  if (!opts.chainId && !opts.providerUrl) {
     // doing a local build, just create a anvil rpc
     node = await runRpc({
       port: 8545,
     });
 
     provider = getProvider(node);
-    chainId = (await provider.getNetwork()).chainId;
   } else {
-    const p = await resolveWriteProvider(cliSettings, opts.chainId);
-    chainId = (await p.provider.getNetwork()).chainId;
+    if (opts.providerUrl && !opts.chainId) {
+      const _provider = new ethers.providers.JsonRpcProvider(opts.providerUrl);
+      chainId = (await _provider.getNetwork()).chainId;
+    }
+    const p = await resolveWriteProvider(cliSettings, chainId || opts.chainId);
+    if (!chainId) {
```

Comment:
this code here is kind of unusual... why not just add an `else` to the if above and have the statement set `chainId = opts.chainId`? Then you can remove `chainId || opts.chainId` from line 161

Reply:
Good point, I though first it's mandatory to resolve the chainId from provider. But you say we can rely on what user has set.